### PR TITLE
Put user id inside session + auth middleware

### PR DESCRIPTION
**Pull request by JiashuHarryHuang**

### CODE REVIEW

File: pages/api/test.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        if (
+          error.message.includes(
+            'Cannot set headers after they are sent to the client'
+          )
+        ) {
+          console.log('User not authenticated');
+        }
+      }
+  }
+}
```
Comment (zineanteoh) [inline pages/api/test.ts:38]: what's this file for?

File: lib/auth.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  try {
+    const session = await getSession({ req });
+    if (!session) {
+      res.status(401).json({ message: 'Not authenticated' });
+    }
+  } catch (err) {
+    console.error(err);
+    return null;
+  }
+}
```
Comment (zineanteoh) [inline lib/auth.ts:17]: is this only used from `test.ts`?

**Review by zineanteoh — CHANGES_REQUESTED** [review]

File: pages/api/test.ts (hunk 1 above)
Comment (JiashuHarryHuang) [reply pages/api/test.ts:38]: Only for testing the middleware

File: lib/auth.ts (hunk 2 above)
Comment (JiashuHarryHuang) [reply lib/auth.ts:17]: yes

**Comment by JiashuHarryHuang** [comment]
Put user id inside session. Implemented auth middleware. For now, auth middleware is called inside an API handler function, which will throw an exception since when the user is not authenticated, the middleware will send a response back, but the handler API handler will continue to execute and attempt to send a response back. So we need to find a way to intercept the request before going to the API handler.

File: lib/getServerSideProps.ts [hunk 3]
```diff
@@ -18,7 +20,7 @@ export async function getServerSideProps(context: GetSessionParams) {
   // If the user is logged in, simply return the session.
   return {
     props: {
-      session,
+      session: JSON.parse(JSON.stringify(session)),
```
Comment (zineanteoh) [inline lib/getServerSideProps.ts:23]: does passing `session` directly not work?

File: lib/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export { default } from 'next-auth/middleware';
+
+export const config = {
+  matcher: ['/api/:function*'],
+};
```
Comment (zineanteoh) [inline lib/middleware.ts:5]: can you add a jsdoc for why this middleware is necessary and what it does?

**Review by zineanteoh — CHANGES_REQUESTED** [review]
Good work on the session + middleware Harry! 👍 
If you can please add some documentation so future us won't forget

File: lib/getServerSideProps.ts (hunk 3 above)
Comment (JiashuHarryHuang) [reply lib/getServerSideProps.ts:23]: There will be a serialization error if passed directly.

File: lib/middleware.ts (hunk 4 above)
Comment (JiashuHarryHuang) [reply lib/middleware.ts:5]: Sure

File: middleware.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+ * to redirect them to the sign-in page.
+ */
+export { default } from 'next-auth/middleware';
+
+/**
+ * Configure which api routes to authenticate
+ */
+export const config = {
+  matcher: ['/api/:function*'],
+};
```
Comment (zineanteoh) [inline middleware.ts:14]: just need to add env variables when we use next auth middleware: https://next-auth.js.org/configuration/nextjs#prerequisites

we can add it in our env template

**Review by zineanteoh — APPROVED** [review]
LGTM!

**Pull request merged**